### PR TITLE
Make Qwen3 device top-k selection exact

### PR DESCRIPTION
**Pull request by zmnobug**
## Summary

Follow up on #769 by replacing the Qwen3-14B per-chunk Top-4 approximation with a provably exact full-vocabulary Top-32 selector.

## What Changed

- Split the 151936 real-vocabulary logits into 74 full 2048-token groups and one 384-token tail group.
- Compute an exact Top-32 for each group with `sort32` and `mrgsort`.
- Merge the resulting `75 x 32 = 2400` candidates in a 4096-entry padded buffer and select the exact global Top-32.
- Keep padded-vocabulary entries excluded with valid-shape masking and minimum-value fill.
- Change the golden implementation to compare directly against `torch.topk(logits[:, :REAL_VOCAB], 32)`.
- Add adversarial fixtures with all Top-32 values concentrated in one 512-token region and distributed across 32 regions.

If an entry is not in its group's Top-32, at least 32 entries in that group rank ahead of it, so it cannot belong to the global Top-32. Therefore the union of exact group Top-32 sets contains the exact global Top-32.

## Validation

- A2/A3 NPU golden comparison: passed for values and token IDs.
- Greedy `selection_k=1` regression: passed.
- Standalone selector, 100 rounds: min 427.1 us, median 427.8 us, mean 428.0 us, max 429.9 us.
- Serving sampling tests after rebase: 29 passed.
- Shared-L3 batch-16 prefill/decode warmup: passed.
- Qwen3-14B 20-token top-k generation: 0.769 s E2E, 26.01 tok/s.

Paired serving PR: https://github.com/hw-native-sys/pypto-serving/pull/77

### CODE REVIEW

File: models/qwen3/14b/topk_select.py [hunk 1]
```diff
@@ -131,68 +154,81 @@ def topk_select_fwd(
...
+                    tail_vals = pl.gather(tail_pairs, mask_pattern=pl.tile.MaskPattern.P0101)
+                    tail_ids = pl.gather(
+                        tail_pairs,
+                        mask_pattern=pl.tile.MaskPattern.P1010,
+                        output_dtype=pl.INT32,
+                    )
+                    tail_offset = TOPK_NUM_FULL_GROUPS * TOPK
+                    for k in pl.range(TOPK):
+                        pl.write(candidate_vals, [0, tail_offset + k], pl.read(tail_vals, [0, k]))
+                        pl.write(candidate_ids, [0, tail_offset + k], pl.read(tail_ids, [0, k]))
```
[the remark this answers is not in the document]
Comment (zmnobug) [reply models/qwen3/14b/topk_select.py]: Addressed in bc73c77: the tail-group path is now guarded by the compile-time condition `TOPK_GROUP_TAIL != 0`, so zero-tail configurations cannot form an out-of-bounds slice. The current Qwen3 exact Top-32 NPU golden test passes.

File: models/qwen3/14b/topk_select.py [hunk 2]
```diff
@@ -25,22 +25,45 @@
...
 assert VOCAB % VOCAB_CHUNK == 0
 assert REAL_VOCAB <= VOCAB
 assert TOPK <= VOCAB_CHUNK
-assert CHUNK_TOPK <= TOPK
-assert CHUNK_TOPK <= CHUNK_TOPK_PAD
-assert REAL_NUM_VOCAB_CHUNKS * CHUNK_TOPK <= CANDIDATE_PAD
+assert TOPK_GROUP_WIDTH % VOCAB_CHUNK == 0
+assert TOPK_GROUP_TAIL == REAL_VOCAB_TAIL
+assert TOPK_NUM_GROUPS * TOPK <= TOPK_CANDIDATE_PAD
 assert NUM_VOCAB_CHUNKS <= GREEDY_CHUNK_PAD
```
[the remark this answers is not in the document]
Comment (zmnobug) [reply models/qwen3/14b/topk_select.py:45]: Addressed in bc73c77: added `assert VOCAB_CHUNK == 512` to make the sort-pipeline invariant explicit and fail fast if the tiling configuration changes.

**Comment by zmnobug** [comment]
Addressed the latest review feedback in bc73c77. In addition to the tail guard and 512-chunk invariant, the distributed Top-32 fixture now spans the full real vocabulary and places candidates in both halves of the final merge. The updated fixture passes exact NPU comparison against `torch.topk`.
